runtime/near-vm-runner/src/wasmtime_runner.rs

```diff
@@ -449,39 +463,31 @@ impl crate::PreparedContract for VMResult<PreparedContract> {
             }
         };
 
-        let engine = module.engine();
-
-        let config = Arc::clone(&result_state.config);
-
-        let ctx = Ctx::default();
-
-        // Clone the caller for future access outside of the store's context
-        let caller = Arc::clone(&ctx.caller);
-        let mut store = Store::<Ctx>::new(engine, ctx);
-        let mut memory = WasmtimeMemory::new(
-            caller,
-            &mut store,
-            config.limit_config.initial_memory_pages,
-            config.limit_config.max_memory_pages,
-        )
-        .expect("failed to construct wasmtime memory");
+        let caller = Arc::default();
+        let mut memory = WasmtimeMemory { caller: Arc::clone(&caller), memory };
         let logic = VMLogic::new(ext, context, fees_config, result_state, &mut memory);
         // SAFETY:
         // Although the 'static here is a lie, we are pretty confident that the `VMLogic` here
         // only lives for the duration of the contract method call (which is covered by the original
         // lifetime).
-        store
-            .data_mut()
-            .logic
-            .replace(unsafe { transmute::<VMLogic<'_>, VMLogic<'static>>(logic) });
-
-        let mut linker = Linker::new(engine);
-        // TODO: config could be accessed through `logic.result_state`, without this code having to
-        // figure it out...
-        link(&mut linker, memory.memory, &store, &config);
-        let res = instantiate_and_call(&mut store, &linker, &module, &method);
+        let logic = unsafe { transmute::<VMLogic<'_>, VMLogic<'static>>(logic) };
+        let ctx = Ctx::new(logic, caller);
+
+        let mut store = Store::<Ctx>::new(pre.module().engine(), ctx);
+        store.limiter(|ctx| &mut ctx.limits);
+        let instance = match pre.instantiate(&mut store) {
+            Ok(instance) => instance,
+            Err(err) => {
+                let err = err.into_vm_error()?;
+                let logic = store.data_mut().logic.take().expect("logic missing");
+                return Ok(VMOutcome::abort(logic.result_state, err));
+            }
+        };
+
+        let res = call(&mut store, instance, &method);
         let logic = store.data_mut().logic.take().expect("logic missing");
-        lazy_drop(Box::new((linker, module)));
+        drop(store);
+        lazy_drop(Box::new(instance));
```

Comment:
Can you add this for now? Was a workaround necessary for wasmer but not sure if wasmtime suffers from the same issue(s) of expensive drop:

```suggestion
        // TODO: verify that lazy dropping actually improves the throughput.
        lazy_drop(Box::new(instance));
```

Reply:
FWIW, https://github.com/near/nearcore/pull/14060 will affect this in significant ways:
- there should not be an expensive drop, since all/most resources should be pooled internally in Wasmtime
- delaying the drop would mean delaying the release of the execution slot (refs the whole semaphore discussion https://github.com/near/nearcore/pull/14060#issuecomment-3183731514)